feat: added thanos check rules command

FUSAKLA opened the pull request:
Resolves #1096 
This adds `thanos check rules` command to check Prometheus rules.

We cannot use the `promtool check rules` since Thanos-rule allows the additional field `partial_response_strategy`

## Changes

- [x] added the check rules cmd
- [x] changelog entry TBD
- [x] docs update?

## Verification

Tested it on different types of invalid rule files even valid. Works fine.

povilasv commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+	}
+}
+
+func checkRulesFiles(files *[]string) error {
```
could you add tests for this.

povilasv commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+func checkRulesFiles(files *[]string) error {
+	var failed error = nil
```
```suggestion
	var failed error
```

povilasv commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+				fmt.Fprintln(os.Stderr, e.Error())
+			}
+			failed = errs[0]
+		} else {
```
remove else statement, and add `continue` to `errs != nil` case.

as per https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

FUSAKLA replied:
hmm this was just copied from the `promtool` but there is still the `fmt.Println()` which is common for both flows. I'd need to duplicate it not sure if that is better.

FUSAKLA replied:
refactored, should be ok now

povilasv commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+			for _, e := range errs {
+				fmt.Fprintln(os.Stderr, e.Error())
+			}
+			failed = errs[0]
```
Maybe we should just join together the errors, instead of returining first error?

FUSAKLA replied:
Yes, I didn't want to over-complicate it since those errors are all printed out nevertheless but in case of reuse of the function in the future it would be better. Will merge those thanks for pointing out.

FUSAKLA commented:
Just one general thought. Should we run this check in the thanos-rule also? 
Would it be more safe to validate the rules before startup to provide better output and and avoid crashing on invalid data? Possibly as a new PR.

bwplotka commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+func checkRules(filename string) (int, []error) {
+	fmt.Println("Checking", filename)
```
some logger would be nice?

bwplotka commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+		return 0, []error{err}
+	}
+
+	// We need to convert Thanos rules to Prometheus rules so we can use their validation
```
missing trailing comma (:

FUSAKLA replied:
thanks, added :)

bwplotka commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+			}
+			failed = fmt.Errorf("Errors: %v", errs)
+		} else {
+			fmt.Printf("  SUCCESS: %d rules found\n", n)
```
can we use logger?

FUSAKLA replied:
Switched to logger everywhere in the check cmd

bwplotka commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,113 @@
+	return nil
+}
+
+func checkRules(filename string) (int, []error) {
```
worth to just return `error` and use  `tsdb.Multierrror`?

FUSAKLA replied:
Great, didn't know about the tsdb multierror. Will change thanks.

FUSAKLA replied:
done, switched to multierror (just a heads up in new version of TSDB multierror moves from generic tsdb package to tsdb/errors package. Imported if according to godoc and then gomod completely broke the dependencies since it upgraded tsdb to new version since the old did not had the package :D )

FUSAKLA commented:
@povilasv Added tests and resolved all your comments PTAL if that's ok with you.

FUSAKLA commented:
So I managed to update even the docs and the genflags script.
Please check those also.

Also the output is now using logger so it looks like this:
```bash
$ ./thanos check rules cmd/thanos/testdata/rules-files/*.yaml ; echo "$?"
level=info ts=2019-05-02T05:33:04.821854781Z caller=check.go:61 msg=Checking filename=cmd/thanos/testdata/rules-files/invalid-rules-data.yaml
level=error ts=2019-05-02T05:33:04.82280852Z caller=check.go:44 result=FAILED
level=error ts=2019-05-02T05:33:04.822879215Z caller=check.go:46 Groupnameshouldnotbeempty=(MISSING)
level=info ts=2019-05-02T05:33:04.822932943Z caller=check.go:48
level=info ts=2019-05-02T05:33:04.822970322Z caller=check.go:61 msg=Checking filename=cmd/thanos/testdata/rules-files/invalid-yaml-format.yaml
level=error ts=2019-05-02T05:33:04.823160151Z caller=check.go:44 result=FAILED
level=error ts=2019-05-02T05:33:04.823198963Z caller=check.go:46 yaml:line4:couldnotfindexpected':'=(MISSING)
level=info ts=2019-05-02T05:33:04.823230325Z caller=check.go:48
level=info ts=2019-05-02T05:33:04.823263265Z caller=check.go:61 msg=Checking filename=cmd/thanos/testdata/rules-files/valid.yaml
level=info ts=2019-05-02T05:33:04.823943932Z caller=check.go:52 result=SUCCESS rulesfound=2
check rules command failed: yaml: line 4: could not find expected ':'
1
```

FUSAKLA commented:
The netify preview fails with
```
7:28:17 AM: Error: Error building site: failed to render pages: render of "page" failed: execute of template failed: html/template:proposal/single.html:5:28: no such template "_default/sidemenu.html"
``` 

Is that me or some flaky behavior?

povilasv commented on docs/components/check.md:
```diff
@@ -0,0 +1,82 @@
+Thanos rule has extended rules file syntax for example with `partial_response_strategy`
+filed which `promtool` does not allow.
+
+If the check fails the command ends with error code `1` otherwise `0`.
```
```suggestion
If the check fails the command fails with exit code `1`, otherwise `0`.
```

FUSAKLA replied:
sorry should be there again

FUSAKLA commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,121 @@
+		level.Info(logger).Log("result", "SUCCESS", "rules found", n)
+	}
+	if failed.Err() != nil {
+		return errors.New(failed.Error())
```
oh it implements the error interface, great thanks

FUSAKLA commented:
Thanks @povilasv ! all applied.

povilasv commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,121 @@
+		level.Info(logger).Log("result", "SUCCESS", "rules found", n)
+	}
+	if failed.Err() != nil {
+		return errors.New(failed.Error())
```
```suggestion
		return failed
```

I think you forgot to do this one :P 
Or do you disagree?

FUSAKLA replied:
sorry should be there again

povilasv commented:
@FUSAKLA you just forced pushed which removed the fixes :P could reapply again?

FUSAKLA commented:
Oh my.. sorry it's been a long day. Thanks for noticing!

![facepalm](https://media.giphy.com/media/11XiZTEwZe15Mk/giphy.gif)

GiedriusS commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func checkRules(logger log.Logger, filename string) (int, tsdb.MultiError) {
+	level.Info(logger).Log("msg", "Checking", "filename", filename)
```
I'm not sure but I think the convention is to start the log messages with a lower-case letter. Could anyone else chime in on this?

FUSAKLA replied:
You are right, I believe it was because of error wrapping maybe? 
Changed to lowercase.

FUSAKLA commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,121 @@
+	return
+}
+
+func registerRulesCheck(m map[string]setupFunc, root *kingpin.CmdClause, name string) {
```
sounds good, renamed

FUSAKLA commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,121 @@
+func registerChecks(m map[string]setupFunc, app *kingpin.Application, name string) {
+	cmd := app.Command(name, "Linting tools for Thanos")
+	registerRulesCheck(m, cmd, name)
+	return
```
removed thanks :)

FUSAKLA commented:
Thanks @GiedriusS, should be fixed and rebased on master now

FUSAKLA commented:
Thanks! Should be up to date now.

GiedriusS commented on cmd/thanos/check.go:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	var rgs thanosrule.RuleGroups
+	if err := yaml.Unmarshal(b, &rgs); err != nil {
```
One more small thing to think about: maybe it would be worth to be strict here too like in `promtool` and use `UnmarshalStrict`? That way we would ensure that no unknown fields are used.

GiedriusS replied:
@FUSAKLA @bwplotka @povilasv what is your opinion on this?

FUSAKLA replied:
Hm, the thing is Thanos rule uses also Unmarshal only. 
Promtool uses strict because Prometheus uses strict. 

Strict can help you with typos for example instead using some defaults and ignoring the misspelled field but I also like the consistency. 

Maybe in next PR make it strict in both cases and use the validation even on the Thanos rule start?

povilasv replied:
IMO, let's do strict here and do a PR to change Thanos RULE to use strict. Nice catch :100:

FUSAKLA replied:
Ok sounds good to me.
Will switch it to the strict mode.

FUSAKLA replied:
Taking it back.. currently this cannot be done because the unmarshaling is done in two steps. 
First unmarshals without strict only the struct with partial response and in second step unmarshals the imported prometheus rule group also with unstrict unmarshall so it passes the additional partial_response_strategy field.

I'll take a closer look but now it fails with the strict.

```go
rs := struct {
		String string `yaml:"partial_response_strategy"`
	}{}

	errMsg := fmt.Sprintf("failed to unmarshal 'partial_response_strategy'. Possible values are %s", strings.Join(storepb.PartialResponseStrategyValues, ","))
	if err := unmarshal(&rs); err != nil {
		return errors.Wrapf(err, errMsg)
	}

	rg := rulefmt.RuleGroup{}
	if err := unmarshal(&rg); err != nil {
		return errors.Wrapf(err, errMsg)
	}
```

povilasv replied:
Ah I see. Maybe we can add our own RuleGroup struct which embedds prometheus one?

```
type RuleGroup struct{
    PartialResponseStrategy string `yaml:"partial_response_strategy"`
    rulefmt.RuleGroup
}
```

This would simplify code, and then we could do unmarshal in one pass + use strict. WDYT?

FUSAKLA replied:
Uh.. looking into it more it's not that simple. `PartialResponseStrategy` is defined in protobuffer and is `int32` the `RuleGroup` defined in `rule.go` has overridden the `UnmarshalYAML` where the mapping from `WARN` to `0` is done. Also setting the default value if the `partial_response_strategy` is missing at all. 

Not sure if it is possible to add own `UnmarshalYAML` to the protobuf type? This could possibly simplify it? 

Maybe @bwplotka could add more info to this I remember we discussed this a bit when he was adding the partial response strategy.

FUSAKLA replied:
Sorry, I didn't notice you changed the type to the `string`. 
This way it can be done but there will be two new structs instead using the Thanos rule ones.

FUSAKLA replied:
Is this what you meant? https://github.com/improbable-eng/thanos/pull/1097/commits/864a74cafb1f301143f11bf5c88cd6333ce1413b

TBH I don't like making the lint and actual loading rules in rule more different but if we eventually unify it in some next PR than I'm OK with it I suppose.

povilasv replied:
Yes, thats what I meant but I think you could do `var rgs []ThanosRuleGroup` instead of `var rgs ThanosRuleGroups`, no ? so then it's only 1 struct :)

FUSAKLA replied:
Great, I thing this cannot be done because I need to unmarshal  the `groups: ` key to the slice.
I can't unmarshal it to slice directly. Or maybe I misunderstood?

povilasv replied:
makes sense. all good :+1:

FUSAKLA commented:
Ok sounds good to me.
Will switch it to the strict mode.

GiedriusS commented:
Could you rebase it one more time? Then I'll give it one last review and I think we can merge this. Thank you for this work and the persistence, it's not easy

FUSAKLA commented:
Sorry for the inactivity, I'm back. 

@GiedriusS rebased so PTAL if you find a time :pray:, thanks!

FUSAKLA commented:
@GiedriusS not sure if this is a bug. It looks like the cmds return error and if the whole process returns any it's printed out by the and this is not using the logger.

For example if you run:
```
$ ./thanos rule --log.format=json
{"caller":"flags.go:87","level":"info","msg":"gossip is disabled","ts":"2019-05-27T21:12:00.055497911Z"}
rule command failed: Gossip is disabled and no --query parameter was given.
```
Yep, it's here I believe?
https://github.com/improbable-eng/thanos/blob/34ad98eda29fc4977462e3e4beaa023186fecdf3/cmd/thanos/main.go#L167

I can print it out using the logger if you want but that would affect all other thanos components but it works as expected so it up to you.
```
$ ./thanos rule --log.format=json
{"caller":"flags.go:87","level":"info","msg":"gossip is disabled","ts":"2019-05-27T21:21:19.432487894Z"}
{"caller":"main.go:168","err":"rule command failed: Gossip is disabled and no --query parameter was given.","level":"error","ts":"2019-05-27T21:21:19.432716717Z"}
```

GiedriusS commented:
> @GiedriusS not sure if this is a bug. It looks like the cmds return error and if the whole process returns any it's printed out by the and this is not using the logger.
> 
> For example if you run:
> 
> ```
> $ ./thanos rule --log.format=json
> {"caller":"flags.go:87","level":"info","msg":"gossip is disabled","ts":"2019-05-27T21:12:00.055497911Z"}
> rule command failed: Gossip is disabled and no --query parameter was given.
> ```
> 
> Yep, it's here I believe?
> 
> https://github.com/improbable-eng/thanos/blob/34ad98eda29fc4977462e3e4beaa023186fecdf3/cmd/thanos/main.go#L167
> 
> I can print it out using the logger if you want but that would affect all other thanos components but it works as expected so it up to you.
> 
> ```
> $ ./thanos rule --log.format=json
> {"caller":"flags.go:87","level":"info","msg":"gossip is disabled","ts":"2019-05-27T21:21:19.432487894Z"}
> {"caller":"main.go:168","err":"rule command failed: Gossip is disabled and no --query parameter was given.","level":"error","ts":"2019-05-27T21:21:19.432716717Z"}
> ```

Sorry, I haven't looked into this completely when commenting originally (didn't see that you haven't introduced this yourself) but yes, I think that should be removed and the proper logger should be used. I mean the user asks for a JSON formatted log so it should be printed like that, no? :smile:

FUSAKLA commented:
@GiedriusS sorry for the delay. I changed the log and rebased on the recent master if that is ok with you now? I agree this way it's more consistent :)

FUSAKLA commented:
I rebased it once again, fixed issues with MultiError from TSDB which moved meantime to different package and updated the docs cmd flags which changed also.  CI is green :tada:
